In [35]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/social-media-addiction-vs-relationships/Students Social Media Addiction.csv


In [36]:
import numpy as np
import pandas as pd

In [37]:
df=pd.read_csv("/kaggle/input/social-media-addiction-vs-relationships/Students Social Media Addiction.csv")

In [38]:
df.shape

(705, 13)

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 705 entries, 0 to 704
Data columns (total 13 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Student_ID                    705 non-null    int64  
 1   Age                           705 non-null    int64  
 2   Gender                        705 non-null    object 
 3   Academic_Level                705 non-null    object 
 4   Country                       705 non-null    object 
 5   Avg_Daily_Usage_Hours         705 non-null    float64
 6   Most_Used_Platform            705 non-null    object 
 7   Affects_Academic_Performance  705 non-null    object 
 8   Sleep_Hours_Per_Night         705 non-null    float64
 9   Mental_Health_Score           705 non-null    int64  
 10  Relationship_Status           705 non-null    object 
 11  Conflicts_Over_Social_Media   705 non-null    int64  
 12  Addicted_Score                705 non-null    int64  
dtypes: fl

In [40]:
df.sample(10)

Student_ID  Age  Gender Academic_Level      Country  \
533         534   21    Male       Graduate  New Zealand   
399         400   22    Male       Graduate       Canada   
280         281   19  Female  Undergraduate        Japan   
83           84   22  Female       Graduate        Malta   
429         430   21    Male       Graduate       Canada   
74           75   23    Male       Graduate      Belarus   
571         572   23    Male       Graduate      Denmark   
203         204   22    Male       Graduate        India   
147         148   20  Female  Undergraduate        India   
205         206   21    Male       Graduate     Pakistan   

     Avg_Daily_Usage_Hours Most_Used_Platform Affects_Academic_Performance  \
533                    4.4           Facebook                          Yes   
399                    4.6             TikTok                          Yes   
280                    3.1               LINE                           No   
83                     2.7           LinkedIn                           No   
429                    4.3          Instagram                          Yes   
74                     2.5           LinkedIn                           No   
571                    3.8            Twitter                           No   
203                    4.6           Facebook                           No   
147                    5.0          Instagram                          Yes   
205                    4.0          Instagram                          Yes   

     Sleep_Hours_Per_Night  Mental_Health_Score Relationship_Status  \
533                    7.8                    7     In Relationship   
399                    7.6                    6              Single   
280                    8.5                    8              Single   
83                     7.2                    8     In Relationship   
429                    7.9                    6              Single   
74                     7.3                    8     In Relationship   
571                    7.8                    8     In Relationship   
203                    8.4                    7     In Relationship   
147                    5.0                    6     In Relationship   
205                    7.7                    5     In Relationship   

     Conflicts_Over_Social_Media  Addicted_Score  
533                            3               6  
399                            3               7  
280                            1               3  
83                             1               4  
429                            3               7  
74                             1               4  
571                            2               4  
203                            2               6  
147                            3               7  
205                            3               7

In [41]:
df["Gender"].value_counts()

# Nominal
# Male:0
# Female: 1
gender_dict={"Male":0,
            "Female":1}
df["Gender_encoded"]=df.Gender.map(gender_dict)
df.drop(["Gender"], axis=1, inplace=True)
df.sample(5)

Student_ID  Age Academic_Level      Country  Avg_Daily_Usage_Hours  \
533         534   21       Graduate  New Zealand                    4.4   
406         407   20  Undergraduate          USA                    7.6   
561         562   21       Graduate        China                    3.9   
99          100   21  Undergraduate       Jordan                    5.4   
397         398   21       Graduate       Mexico                    6.3   

    Most_Used_Platform Affects_Academic_Performance  Sleep_Hours_Per_Night  \
533           Facebook                          Yes                    7.8   
406             TikTok                          Yes                    4.7   
561             WeChat                           No                    8.4   
99              TikTok                          Yes                    5.5   
397           WhatsApp                          Yes                    5.9   

     Mental_Health_Score Relationship_Status  Conflicts_Over_Social_Media  \
533                    7     In Relationship                            3   
406                    5     In Relationship                            4   
561                    7     In Relationship                            2   
99                     5     In Relationship                            4   
397                    6              Single                            3   

     Addicted_Score  Gender_encoded  
533               6               0  
406               9               1  
561               5               0  
99                8               1  
397               7               0

In [42]:
df["Academic_Level"].value_counts()

# Ordinal
# Highschool:1
# UG:2, Graduate:3

Academic_Level
Undergraduate    353
Graduate         325
High School       27
Name: count, dtype: int64

In [43]:
acad_level_dict={"High School": 0,
            "Undergraduate": 1,
            "Graduate": 2}
df["acad_level_encoded"]=df.Academic_Level.map(acad_level_dict)
df.drop(["Academic_Level"], axis=1, inplace=True)
df


Student_ID  Age     Country  Avg_Daily_Usage_Hours Most_Used_Platform  \
0             1   19  Bangladesh                    5.2          Instagram   
1             2   22       India                    2.1            Twitter   
2             3   20         USA                    6.0             TikTok   
3             4   18          UK                    3.0            YouTube   
4             5   21      Canada                    4.5           Facebook   
..          ...  ...         ...                    ...                ...   
700         701   20       Italy                    4.7             TikTok   
701         702   23      Russia                    6.8          Instagram   
702         703   21       China                    5.6             WeChat   
703         704   24       Japan                    4.3            Twitter   
704         705   19      Poland                    6.2           Facebook   

    Affects_Academic_Performance  Sleep_Hours_Per_Night  Mental_Health_Score  \
0                            Yes                    6.5                    6   
1                             No                    7.5                    8   
2                            Yes                    5.0                    5   
3                             No                    7.0                    7   
4                            Yes                    6.0                    6   
..                           ...                    ...                  ...   
700                           No                    7.2                    7   
701                          Yes                    5.9                    4   
702                          Yes                    6.7                    6   
703                           No                    7.5                    8   
704                          Yes                    6.3                    5   

    Relationship_Status  Conflicts_Over_Social_Media  Addicted_Score  \
0       In Relationship                            3               8   
1                Single                            0               3   
2           Complicated                            4               9   
3                Single                            1               4   
4       In Relationship                            2               7   
..                  ...                          ...             ...   
700     In Relationship                            2               5   
701              Single                            5               9   
702     In Relationship                            3               7   
703              Single                            2               4   
704              Single                            4               8   

     Gender_encoded  acad_level_encoded  
0                 1                   1  
1                 0                   2  
2                 1                   1  
3                 0                   0  
4                 0                   2  
..              ...                 ...  
700               1                   1  
701               0                   2  
702               1                   1  
703               0                   2  
704               1                   1  

[705 rows x 13 columns]

In [44]:
print(df["Country"].nunique())
df["Country"].value_counts()

# nominal
# top 10 OHE rest in others

110


Country
India        53
USA          40
Canada       34
Denmark      27
Ireland      27
             ..
Uruguay       1
Ecuador       1
Venezuela     1
Peru          1
Jamaica       1
Name: count, Length: 110, dtype: int64

In [46]:
# Step 1: Get top 10 countries
top_10_countries = df['Country'].value_counts().nlargest(10).index

# Step 2: Replace others with 'Other'
df['Country_top10'] = df['Country'].apply(lambda x: x if x in top_10_countries else 'Other')

# Step 3: One-hot encode with 0/1
df_country_dummies = pd.get_dummies(df['Country_top10'], prefix='Country', dtype=int)

# Step 4: Merge with original df
df = pd.concat([df, df_country_dummies], axis=1)

# Step 5: Drop unnecessary columns
df.drop(['Country', 'Country_top10'], axis=1, inplace=True)


In [49]:
df

Student_ID  Age  Avg_Daily_Usage_Hours Most_Used_Platform  \
0             1   19                    5.2          Instagram   
1             2   22                    2.1            Twitter   
2             3   20                    6.0             TikTok   
3             4   18                    3.0            YouTube   
4             5   21                    4.5           Facebook   
..          ...  ...                    ...                ...   
700         701   20                    4.7             TikTok   
701         702   23                    6.8          Instagram   
702         703   21                    5.6             WeChat   
703         704   24                    4.3            Twitter   
704         705   19                    6.2           Facebook   

    Affects_Academic_Performance  Sleep_Hours_Per_Night  Mental_Health_Score  \
0                            Yes                    6.5                    6   
1                             No                    7.5                    8   
2                            Yes                    5.0                    5   
3                             No                    7.0                    7   
4                            Yes                    6.0                    6   
..                           ...                    ...                  ...   
700                           No                    7.2                    7   
701                          Yes                    5.9                    4   
702                          Yes                    6.7                    6   
703                           No                    7.5                    8   
704                          Yes                    6.3                    5   

    Relationship_Status  Conflicts_Over_Social_Media  Addicted_Score  ...  \
0       In Relationship                            3               8  ...   
1                Single                            0               3  ...   
2           Complicated                            4               9  ...   
3                Single                            1               4  ...   
4       In Relationship                            2               7  ...   
..                  ...                          ...             ...  ...   
700     In Relationship                            2               5  ...   
701              Single                            5               9  ...   
702     In Relationship                            3               7  ...   
703              Single                            2               4  ...   
704              Single                            4               8  ...   

     Country_Denmark  Country_France  Country_India  Country_Ireland  \
0                  0               0              0                0   
1                  0               0              1                0   
2                  0               0              0                0   
3                  0               0              0                0   
4                  0               0              0                0   
..               ...             ...            ...              ...   
700                0               0              0                0   
701                0               0              0                0   
702                0               0              0                0   
703                0               0              0                0   
704                0               0              0                0   

     Country_Mexico  Country_Other  Country_Spain  Country_Switzerland  \
0                 0              1              0                    0   
1                 0              0              0                    0   
2                 0              0              0                    0   
3                 0              1              0                    0   
4                 0              0              0                    0   
..              .

In [48]:
df.columns

Index(['Student_ID', 'Age', 'Avg_Daily_Usage_Hours', 'Most_Used_Platform',
       'Affects_Academic_Performance', 'Sleep_Hours_Per_Night',
       'Mental_Health_Score', 'Relationship_Status',
       'Conflicts_Over_Social_Media', 'Addicted_Score', 'Gender_encoded',
       'acad_level_encoded', 'Country_Canada', 'Country_Denmark',
       'Country_France', 'Country_India', 'Country_Ireland', 'Country_Mexico',
       'Country_Other', 'Country_Spain', 'Country_Switzerland',
       'Country_Turkey', 'Country_USA'],
      dtype='object')

In [50]:
df["Most_Used_Platform"].value_counts()

# nominal
# frequncy encoding

Most_Used_Platform
Instagram    249
TikTok       154
Facebook     123
WhatsApp      54
Twitter       30
LinkedIn      21
WeChat        15
Snapchat      13
LINE          12
KakaoTalk     12
VKontakte     12
YouTube       10
Name: count, dtype: int64

In [52]:
fe=df.groupby("Most_Used_Platform").size()/len(df)
df.loc[:, "Most_Used_Platform_encod"] = df["Most_Used_Platform"].map(fe)
df.drop("Most_Used_Platform", axis=1, inplace=True)
df

Student_ID  Age  Avg_Daily_Usage_Hours Affects_Academic_Performance  \
0             1   19                    5.2                          Yes   
1             2   22                    2.1                           No   
2             3   20                    6.0                          Yes   
3             4   18                    3.0                           No   
4             5   21                    4.5                          Yes   
..          ...  ...                    ...                          ...   
700         701   20                    4.7                           No   
701         702   23                    6.8                          Yes   
702         703   21                    5.6                          Yes   
703         704   24                    4.3                           No   
704         705   19                    6.2                          Yes   

     Sleep_Hours_Per_Night  Mental_Health_Score Relationship_Status  \
0                      6.5                    6     In Relationship   
1                      7.5                    8              Single   
2                      5.0                    5         Complicated   
3                      7.0                    7              Single   
4                      6.0                    6     In Relationship   
..                     ...                  ...                 ...   
700                    7.2                    7     In Relationship   
701                    5.9                    4              Single   
702                    6.7                    6     In Relationship   
703                    7.5                    8              Single   
704                    6.3                    5              Single   

     Conflicts_Over_Social_Media  Addicted_Score  Gender_encoded  ...  \
0                              3               8               1  ...   
1                              0               3               0  ...   
2                              4               9               1  ...   
3                              1               4               0  ...   
4                              2               7               0  ...   
..                           ...             ...             ...  ...   
700                            2               5               1  ...   
701                            5               9               0  ...   
702                            3               7               1  ...   
703                            2               4               0  ...   
704                            4               8               1  ...   

     Country_France  Country_India  Country_Ireland  Country_Mexico  \
0                 0              0                0               0   
1                 0              1                0               0   
2                 0              0                0               0   
3                 0              0                0               0   
4                 0              0                0               0   
..              ...            ...              ...             ...   
700               0              0                0               0   
701               0              0                0               0   
702               0              0                0               0   
703               0              0                0               0   
704               0              0                0               0   

     Country_Other  Country_Spain  Country_Switzerland  Country_Turkey  \
0                1              0                    0               0   
1                0              0                    0               0   
2                0              0                    0               0   
3                1              0                    0               0   
4                0              0                    0               0   
..             ...            ...                  ...           

In [53]:
df["Affects_Academic_Performance"].value_counts()

# nominal but can be replaced as Yes:1, No:0

Affects_Academic_Performance
Yes    453
No     252
Name: count, dtype: int64

In [54]:
import category_encoders as ce
encoder = ce.BinaryEncoder(cols=["Affects_Academic_Performance"])
dfbin = encoder.fit_transform(df["Affects_Academic_Performance"])
df=pd.concat([df, dfbin], axis=1)
df

Student_ID  Age  Avg_Daily_Usage_Hours Affects_Academic_Performance  \
0             1   19                    5.2                          Yes   
1             2   22                    2.1                           No   
2             3   20                    6.0                          Yes   
3             4   18                    3.0                           No   
4             5   21                    4.5                          Yes   
..          ...  ...                    ...                          ...   
700         701   20                    4.7                           No   
701         702   23                    6.8                          Yes   
702         703   21                    5.6                          Yes   
703         704   24                    4.3                           No   
704         705   19                    6.2                          Yes   

     Sleep_Hours_Per_Night  Mental_Health_Score Relationship_Status  \
0                      6.5                    6     In Relationship   
1                      7.5                    8              Single   
2                      5.0                    5         Complicated   
3                      7.0                    7              Single   
4                      6.0                    6     In Relationship   
..                     ...                  ...                 ...   
700                    7.2                    7     In Relationship   
701                    5.9                    4              Single   
702                    6.7                    6     In Relationship   
703                    7.5                    8              Single   
704                    6.3                    5              Single   

     Conflicts_Over_Social_Media  Addicted_Score  Gender_encoded  ...  \
0                              3               8               1  ...   
1                              0               3               0  ...   
2                              4               9               1  ...   
3                              1               4               0  ...   
4                              2               7               0  ...   
..                           ...             ...             ...  ...   
700                            2               5               1  ...   
701                            5               9               0  ...   
702                            3               7               1  ...   
703                            2               4               0  ...   
704                            4               8               1  ...   

     Country_Ireland  Country_Mexico  Country_Other  Country_Spain  \
0                  0               0              1              0   
1                  0               0              0              0   
2                  0               0              0              0   
3                  0               0              1              0   
4                  0               0              0              0   
..               ...             ...            ...            ...   
700                0               0              1              0   
701                0               0              1              0   
702                0               0              1              0   
703                0               0              1              0   
704                0               0              1              0   

     Country_Switzerland  Country_Turkey  Country_USA  \
0                      0               0            0   
1                      0               0            0   
2                      0               0            1   
3                      0               0            0   
4                      0               0            0   
..                   ...             ...          ...   
700                    0               0            0   
701                    0               0            0   
702     

In [55]:
df.drop(["Affects_Academic_Performance_1", "Affects_Academic_Performance"], axis=1, inplace=True)
df

Student_ID  Age  Avg_Daily_Usage_Hours  Sleep_Hours_Per_Night  \
0             1   19                    5.2                    6.5   
1             2   22                    2.1                    7.5   
2             3   20                    6.0                    5.0   
3             4   18                    3.0                    7.0   
4             5   21                    4.5                    6.0   
..          ...  ...                    ...                    ...   
700         701   20                    4.7                    7.2   
701         702   23                    6.8                    5.9   
702         703   21                    5.6                    6.7   
703         704   24                    4.3                    7.5   
704         705   19                    6.2                    6.3   

     Mental_Health_Score Relationship_Status  Conflicts_Over_Social_Media  \
0                      6     In Relationship                            3   
1                      8              Single                            0   
2                      5         Complicated                            4   
3                      7              Single                            1   
4                      6     In Relationship                            2   
..                   ...                 ...                          ...   
700                    7     In Relationship                            2   
701                    4              Single                            5   
702                    6     In Relationship                            3   
703                    8              Single                            2   
704                    5              Single                            4   

     Addicted_Score  Gender_encoded  acad_level_encoded  ...  Country_India  \
0                 8               1                   1  ...              0   
1                 3               0                   2  ...              1   
2                 9               1                   1  ...              0   
3                 4               0                   0  ...              0   
4                 7               0                   2  ...              0   
..              ...             ...                 ...  ...            ...   
700               5               1                   1  ...              0   
701               9               0                   2  ...              0   
702               7               1                   1  ...              0   
703               4               0                   2  ...              0   
704               8               1                   1  ...              0   

     Country_Ireland  Country_Mexico  Country_Other  Country_Spain  \
0                  0               0              1              0   
1                  0               0              0              0   
2                  0               0              0              0   
3                  0               0              1              0   
4                  0               0              0              0   
..               ...             ...            ...            ...   
700                0               0              1              0   
701                0               0              1              0   
702                0               0              1              0   
703                0               0              1              0   
704                0               0              1              0   

     Country_Switzerland  Country_Turkey  Country_USA  \
0                      0               0            0   
1                      0               0            0   
2                      0               0            1   
3                      0               0            0   
4                      0               0            0   
..                   ...             ...          ...   
700                    0               0          

In [56]:
df["Relationship_Status"].value_counts()

# nominal
# OHE

Relationship_Status
Single             384
In Relationship    289
Complicated         32
Name: count, dtype: int64

In [58]:
df_Relationship_Status_dummies = pd.get_dummies(df['Relationship_Status'], prefix='Relationship_Status', dtype=int)
df = pd.concat([df, df_Relationship_Status_dummies], axis=1)
df.drop('Relationship_Status', axis=1, inplace=True)
df

Student_ID  Age  Avg_Daily_Usage_Hours  Sleep_Hours_Per_Night  \
0             1   19                    5.2                    6.5   
1             2   22                    2.1                    7.5   
2             3   20                    6.0                    5.0   
3             4   18                    3.0                    7.0   
4             5   21                    4.5                    6.0   
..          ...  ...                    ...                    ...   
700         701   20                    4.7                    7.2   
701         702   23                    6.8                    5.9   
702         703   21                    5.6                    6.7   
703         704   24                    4.3                    7.5   
704         705   19                    6.2                    6.3   

     Mental_Health_Score  Conflicts_Over_Social_Media  Addicted_Score  \
0                      6                            3               8   
1                      8                            0               3   
2                      5                            4               9   
3                      7                            1               4   
4                      6                            2               7   
..                   ...                          ...             ...   
700                    7                            2               5   
701                    4                            5               9   
702                    6                            3               7   
703                    8                            2               4   
704                    5                            4               8   

     Gender_encoded  acad_level_encoded  Country_Canada  ...  Country_Turkey  \
0                 1                   1               0  ...               0   
1                 0                   2               0  ...               0   
2                 1                   1               0  ...               0   
3                 0                   0               0  ...               0   
4                 0                   2               1  ...               0   
..              ...                 ...             ...  ...             ...   
700               1                   1               0  ...               0   
701               0                   2               0  ...               0   
702               1                   1               0  ...               0   
703               0                   2               0  ...               0   
704               1                   1               0  ...               0   

     Country_USA  Most_Used_Platform_encod  Affects_Academic_Performance_0  \
0              0                  0.353191                               0   
1              0                  0.042553                               1   
2              1                  0.218440                               0   
3              0                  0.014184                               1   
4              0                  0.174468                               0   
..           ...                       ...                             ...   
700            0                  0.218440                               1   
701            0                  0.353191                               0   
702            0                  0.021277                               0   
703            0                  0.042553                               1   
704            0                  0.174468                               0   

     Relationship_Status_Complicated  Relationship_Status_In Relationship  \
0                                  0                                    1   
1                                  0                                    0   
2                                  1                                    0   
3                                  0                                    0   
4   

In [61]:
from sklearn.model_selection import train_test_split

In [62]:
X = df.drop('Addicted_Score', axis=1)
y = df['Addicted_Score'] 

In [63]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [64]:
from sklearn.linear_model import LinearRegression

lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

LinearRegression()

In [65]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred = lr_model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R² Score: {r2}")

Mean Squared Error: 0.08163073607223176
R² Score: 0.9673775696741168


In [68]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
poly = PolynomialFeatures(degree=2)

X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

model = LinearRegression()
model.fit(X_train_poly, y_train)

LinearRegression()

In [69]:
y_pred = model.predict(X_test_poly)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"MSE: {mse}")
print(f"R²: {r2}")

MSE: 891786606334.7346
R²: -356388387884.76135


In [ ]:
# R² is negative interprets "chullu bhar pani naak me bhar lo"

In [70]:
# Tree based models

from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor

In [71]:
models = {
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "Random Forest": RandomForestRegressor(random_state=42),
    "Gradient Boosting": GradientBoostingRegressor(random_state=42)
}

In [72]:
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)

    print(f"\n {name}")
    print(f"Mean Squared Error: {mse:.4f}")
    print(f"R² Score: {r2:.4f}")


 Decision Tree
Mean Squared Error: 0.0355
R² Score: 0.9858

 Random Forest
Mean Squared Error: 0.0296
R² Score: 0.9882

 Gradient Boosting
Mean Squared Error: 0.0336
R² Score: 0.9866


In [ ]:
# all the scores are are almost similar

In [73]:
# Support Vector Regression

from sklearn.svm import SVR

In [75]:
from sklearn.preprocessing import StandardScaler
# Scale both X and y very important for SVR
scaler_X = StandardScaler()
scaler_y = StandardScaler()

X_train_scaled = scaler_X.fit_transform(X_train)
X_test_scaled = scaler_X.transform(X_test)

# y needs to be reshaped before scaling
y_train_scaled = scaler_y.fit_transform(y_train.values.reshape(-1, 1)).ravel()

In [76]:
svr_model = SVR(kernel='rbf')  # 'rbf' is the default; can try 'linear' or 'poly'
svr_model.fit(X_train_scaled, y_train_scaled)


SVR()

In [77]:
y_pred_scaled = svr_model.predict(X_test_scaled)

# Inverse transform to get predictions in original scale
y_pred = scaler_y.inverse_transform(y_pred_scaled.reshape(-1, 1)).ravel()
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R² Score: {r2}")

Mean Squared Error: 0.06670059830092824
R² Score: 0.9733441626835098


In [ ]:
# reduced score

In [78]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'C': [0.1, 1, 10],
    'epsilon': [0.1, 0.2, 0.5],
    'kernel': ['linear', 'rbf']
}

grid = GridSearchCV(SVR(), param_grid, cv=5)
grid.fit(X_train_scaled, y_train_scaled)

print("Best Parameters:", grid.best_params_)


Best Parameters: {'C': 10, 'epsilon': 0.1, 'kernel': 'rbf'}


In [79]:
svr_model = SVR(C=10, epsilon=0.1, kernel='rbf')
svr_model.fit(X_train_scaled, y_train_scaled)
y_pred_scaled = svr_model.predict(X_test_scaled)
y_pred = scaler_y.inverse_transform(y_pred_scaled.reshape(-1, 1)).ravel()

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"SVR Performance:")
print(f"Mean Squared Error: {mse}")
print(f"R² Score: {r2}")

SVR Performance:
Mean Squared Error: 0.0568492708482791
R² Score: 0.9772810896169768


In [ ]:
# no improvement

In [80]:
from sklearn.neighbors import KNeighborsRegressor

In [81]:
knn = KNeighborsRegressor(n_neighbors=5)
knn.fit(X_train, y_train)

KNeighborsRegressor()

In [82]:
y_pred = knn.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse:.4f}")
print(f"R² Score: {r2:.4f}")


Mean Squared Error: 1.6167
R² Score: 0.3539


In [ ]:
# not at all worth it

In [89]:
knn = KNeighborsRegressor(n_neighbors=2)  # try 3, 5, 7, etc.
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse:.4f}")
print(f"R² Score: {r2:.4f}")


Mean Squared Error: 1.4309
R² Score: 0.4282


In [ ]:
# only highest score we can get

In [ ]:
# Now applying the neural network regression

In [90]:
from sklearn.neural_network import MLPRegressor
mlp = MLPRegressor(hidden_layer_sizes=(100, ),   # One hidden layer with 100 neurons
                   activation='relu',             # Activation function
                   solver='adam',                 # Optimization algorithm
                   max_iter=500,                  # Max iterations for convergence
                   random_state=42)

mlp.fit(X_train, y_train)

MLPRegressor(max_iter=500, random_state=42)

In [91]:
y_pred = mlp.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

Mean Squared Error: 0.10442357696937964
R^2 Score: 0.9582687719359927


In [95]:
# trying with scaled values
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

mlp = MLPRegressor(hidden_layer_sizes=(100, 100), activation='relu', max_iter=1000, random_state=42)
mlp.fit(X_train, y_train)

MLPRegressor(hidden_layer_sizes=(100, 100), max_iter=1000, random_state=42)

In [96]:
y_pred = mlp.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

Mean Squared Error: 0.12311260450212357
R^2 Score: 0.9507999981887368


In [ ]:
# hahahahahahahhahahahahahahahahahah

1. Linear Regression
MSE: 0.0816
R²: 0.9674

Reason for Good Performance:
The data has a strong linear relationship.
Minimal multicollinearity or noise.
The target is well-predicted using a linear combination of features.

2. Polynomial Regression (with Linear Regression)
MSE: 891,786,606,334.7346
R²: -356,388,387,884.76135

Reason for Catastrophic Performance:
Polynomial features overfit massively or exploded feature values (especially if not scaled).
If the degree is high (e.g., 3 or 4), and features aren't normalized, the model fails.
Likely the target values are small, but features became huge after transformation.

3. Decision Tree Regressor
MSE: 0.0355
R²: 0.9858

Reason for Good Performance:
Decision trees handle nonlinear relationships and variable interactions well.
Performs well with little preprocessing.
But can overfit if not controlled (use max_depth, min_samples_leaf, etc.).

4. Random Forest Regressor
MSE: 0.0296
R²: 0.9882

Best performance among all so far!
Reason:
Random Forest is an ensemble of trees — reduces overfitting and variance.
Robust to noise and outliers.
Works well even with unscaled data.

5. Gradient Boosting Regressor
MSE: 0.0336
R²: 0.9866

Reason for Good Performance:
Boosts model performance iteratively by focusing on errors.
Typically outperforms random forest on clean, structured data with proper tuning.
Slightly behind Random Forest in your case, possibly due to default parameters.

6. SVR (Support Vector Regression)
MSE: 0.0667
R²: 0.9733

Reason:
SVR works well on smaller datasets with high-dimensional features.
Sensitive to feature scaling — performance is good if scaled properly.
May not model complex nonlinearities as well as tree models.

7. SVR with GridSearchCV
MSE: 0.0568
R²: 0.9773

Reason:
Hyperparameter tuning improved SVR performance.
Better C, epsilon, or kernel selected.
Still behind ensemble tree models but significantly improved.

8. KNN Regressor
MSE: 1.6167
R²: 0.3539

KNN (k=2):
MSE: 1.4309
R²: 0.4282

Reason for Poor Performance:
KNN is very sensitive to noise and feature scaling.
Performs poorly on high-dimensional data unless well-tuned.
Also, not good for extrapolation — only interpolates.

9. MLPRegressor (Neural Network)
MSE: 0.1231
R²: 0.9508

Reason for Slightly Lower Performance:
Neural networks need a lot of data and scaling to perform best.
You likely used StandardScaler, which helped.
Could still benefit from hyperparameter tuning (layers, learning rate, etc.).
But generally performs well and is more flexible than linear models.


Summary of Model Rankings (based on R²):
Model	                R² Score	  Verdict
Random Forest	        0.9882	      Excellent
Gradient Boosting	    0.9866	      Excellent
Decision Tree	        0.9858	      Very Good
Linear Regression	    0.9674	      Very Good
SVR (Tuned)	            0.9773	      Very Good
SVR (Default)	        0.9733	      Good
MLPRegressor	        0.9508	      Good (needs tuning)
KNN	                    0.35–0.42	  Poor (needs tuning)
Polynomial Regression	-356B	      Completely Failed